modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/richtext/skin/RichTextAreaSkin.java

```diff
@@ -192,6 +235,142 @@ public void dispose() {
+        double maxY = 0.0;
+
+        List<PathElement> elements = vflow.getUnderlineShape(start, end);
+        int sz = elements.size();
+        for (int i = 0; i < sz; i++) {
+            PathElement pe = elements.get(i);
+            if (pe instanceof MoveTo em) {
+                minX = maxX = em.getX();
+                minY = maxY = em.getY();
+            } else if (pe instanceof LineTo em) {
+                minX = (minX < em.getX() ? minX : em.getX());
+                maxX = (maxX > em.getX() ? maxX : em.getX());
+                minY = (minY < em.getY() ? minY : em.getY());
+                maxY = (maxY > em.getY() ? maxY : em.getY());
+            } else if (pe instanceof HLineTo em) {
+                minX = (minX < em.getX() ? minX : em.getX());
+                maxX = (maxX > em.getX() ? maxX : em.getX());
+            } else if (pe instanceof VLineTo em) {
+                minY = (minY < em.getY() ? minY : em.getY());
+                maxY = (maxY > em.getY() ? maxY : em.getY());
+            }
+            // don't assume that shapes are ended with ClosePath
+            if (
+                pe instanceof ClosePath ||
+                i == sz - 1 ||
+                (i < sz - 1 && elements.get(i + 1) instanceof MoveTo)
+            )
+            {
+                // create the shape
+                Shape sh = null;
+                switch(highlight) {
+                case SELECTED_RAW:
+                    // blue background
+                    sh = new Path(vflow.getRangeShape(start, end));
+                    sh.setFill(imeSelectColor());
+                    sh.setOpacity(0.3);
+                    break;
+                case UNSELECTED_RAW:
+                    // dash underline
+                    sh = new Line(minX + 2, maxY + 1, maxX - 2, maxY + 1);
+                    sh.setStroke(imeColor());
+                    sh.setStrokeWidth(maxY - minY);
+                    ObservableList<Double> dashArray = sh.getStrokeDashArray();
+                    dashArray.add(2.0);
+                    dashArray.add(2.0);
+                    break;
+                case SELECTED_CONVERTED:
+                    // thick underline
+                    sh = new Line(minX + 2, maxY + 1, maxX - 2, maxY + 1);
+                    sh.setStroke(imeColor());
+                    sh.setStrokeWidth((maxY - minY) * 3);
+                    break;
+                case UNSELECTED_CONVERTED:
+                    // single underline
+                    sh = new Line(minX + 2, maxY + 1, maxX - 2, maxY + 1);
+                    sh.setStroke(imeColor());
+                    sh.setStrokeWidth(maxY - minY);
+                    break;
+                }
+
+                if (sh != null) {
+                    sh.setManaged(false);
+                    shapes.add(sh);
+                }
+            }
+        }
+    }
+
+    private Color imeColor() {
+        return RichUtils.isDarkScheme(getSkinnable()) ? Color.WHITE : Color.BLACK;
```

Comment:
I don't see anything similar in TextInputControlSkin. Should there be (as a follow-up bug)?